src/controllers/tags.js

```diff
@@ -0,0 +1,76 @@
+// APLICATION LOGIC FOR TAGGING
+const { catchAsync, AppError } = require('../lib')
+const tagService = require('../services/tags')
+
+/**
+ * Destructure tag 
+ * @param {object} tag - The tag object
+ */
+const transformTag = (tag) => {
+	const data = {
+		id: tag.id,
+		title: tag.title,
+		content: tag.content
+	}
+	return data
+}
+
+/** Controller for creating a new tag */ 
+const createNewTag = catchAsync(async (req, res, next) => {
+	const { content, title } = req.body
+	const newTag = await tagService.createTag({ content, title })
+
+	if (!newTag) return next(new AppError('Tag already exists', 400))
+	const data = transformTag(newTag)
+
+	return res.status(201).json({
+		status: 'success',
+		data
+	})
+})
+
+/** Controller for fetching all tags */ 
+const fetchTags = catchAsync(async (req, res) => {
+	const allTags = await tagService.fetchTags()
+
+	return res.status(200).json({
+		status: 'success',
+		data: allTags.map(transformTag)
+	})
+})
+
+/** Controller for updating a tag */ 
+const updateTag = catchAsync(async (req, res, next) => {
+	const { tagId } = req.params
+	const { content, title } = req.body
+
+	const updatedTag = await tagService.updateTag(title, content, tagId)
+
+	if (!updatedTag) {
+		return next(new AppError('Tag does not exist', 404))
```

Comment:
@itzgeebee this is the final thing i've been thinking about then we're done :smile: . I like the use of `AppError` especially and the `catchAsync` too.

so for this refactoring phase, we want to have our errors in one place and also not hardcoded for reuse and hopefully more readability. 

so for now 
1. Check the `services/errors.js`, add a `TagDoesNotExistError`
2. Move this check into the tag service and throw the `TagDoesNotExistError` you've created. Example in `services/posts.js`.
3. Have an error handler for the route as a whole in `routes/tags.js` so that we will do the mapping between the [error and the status code](https://github.com/tbf-challenge/teamwork-be/blob/main/src/routes/articles.js#L160-L170). Example in [routes/articles.js](https://github.com/tbf-challenge/teamwork-be/blob/main/src/routes/articles.js#L160-L170) Helps to not look around looking for which error returns which status code. 

Kindly let me know if anything and I can reach out on our Slack too if there's more clarification you might want to get there

Reply:
@jocrah Just to clarify. I shouldn't use the `catchAsync` and `AppError` and use the try-catch one in the posts and articles?

Reply:
not at all @itzgeebee. I was actually saying like the usage and it's likely we'll eventually change the others to that. 

The main part is that we want the errors a bit centralized hence the 3 steps I gave.